Add default-flags configurations for all TFLite benchmarks

GMNGeoffrey opened the pull request:
Following up on https://github.com/google/iree/pull/7553, this adds
default flag configurations for the remaining TFLite models. Now that
we've got timestamps in the Buildkite runs, we can see what this does
to artifact transfer time via the presubmit:
https://buildkite.com/iree/iree-benchmark/builds/1447

I'm contrasting with the run at the merge base on main:
https://buildkite.com/iree/iree-benchmark/builds/1445

For this PR's run, downloads took 1:35 and 4:12 for the two RPIs,
compared to 2:06 and 3:30 before. That looks like it's basically in
the noise. We could do more detailed analysis, but that doesn't seem
worth it to me. Just spot-checking a few, it seems the RPI connected
to the Pixels consistently downloads faster than the ones connected
to the Samsungs.

As for the benchmarks themselves, the run on the RPI took
around 34-35 minutes for each phone, 6-8 min longer than
before.

Overall run latency, excluding waiting for agents, increased
from 43:12 to 49:50. I think this is acceptable given our
current limitations.

antiagainst commented on benchmarks/TFLite/CMakeLists.txt:
```diff
@@ -60,14 +60,20 @@ set(POSENET_FP32_MODULE
+# Each suite benchmarks a list of modules with configurations specifying a     #
+# target architecture and runtime characteristics (e.g. threads/cores). These  #
+# benchmarks only configure IREE translation and runtime flags for these       #
+# purposes target architecture and do *not* include any non-default flags. No  #
```
..  for these ~purposes~ target ..

antiagainst commented on benchmarks/TFLite/CMakeLists.txt:
```diff
@@ -60,14 +60,20 @@ set(POSENET_FP32_MODULE
 #                                                                              #
 ################################################################################
 
+set(ANDROID_CPU_TRANSLATION_FLAGS
```
`_CPU_DEFAULT_ `?

antiagainst replied:
I see. That makes sense then.

antiagainst commented on benchmarks/TFLite/CMakeLists.txt:
```diff
@@ -109,14 +109,134 @@ iree_benchmark_suite(
+)
+
+
+# GPU, Vulkan, Mali, full-inference
```
Missing configuration for Adreno GPUs?

antiagainst replied:
Yeah. It's fine to omit them in the default settings for now.

antiagainst commented on benchmarks/TFLite/CMakeLists.txt:
```diff
@@ -76,25 +82,19 @@ iree_benchmark_suite(
-    "--iree-llvm-target-triple=aarch64-none-linux-android29"
-    "--iree-flow-inline-constants-max-byte-length=2048"
-    "--iree-llvm-loop-unrolling=true"
+    "${ANDROID_CPU_TRANSLATION_FLAGS}"
```
This should not quite given we expect a list here?

antiagainst replied:
Yeah, quoted. Sorry about the typo! ;-P

antiagainst commented on benchmarks/TFLite/CMakeLists.txt:
```diff
@@ -109,14 +109,134 @@ iree_benchmark_suite(
     "--task_topology_group_count=1"
 )
 
+# CPU, Dylib, 2 through 4-thread, little-core, full-inference
```
Hmm, these makes me wonder we should have another level of permutation for `RUNTIME_FLAGS` to avoid these explicit copies.. Gah. CMake is really not good at this..

(No request for change for now.)

antiagainst reviewed (changes requested):
The change LGTM. Putting a blocker here mostly wanting to see how we make the change given you updated existing benchmarks to add `experimental-flags` for them. I'm thinking to stop the current sever, do a database find/replace, bring it back on online, and then land this pull request. This would mean to do it in a pause-the-world way. I think that should be fine: we can explicitly ask everybody to not land, or I can do it (given the timezone difference when I'm working everybody is still asleep). Anyway, we need to see how to migrate over existing series; so need to do some experiments. WDYT?

GMNGeoffrey commented:
> The change LGTM. Putting a blocker here mostly wanting to see how we make the change given you updated existing benchmarks to add `experimental-flags` for them. I'm thinking to stop the current sever, do a database find/replace, bring it back on online, and then land this pull request. This would mean to do it in a pause-the-world way. I think that should be fine: we can explicitly ask everybody to not land, or I can do it (given the timezone difference when I'm working everybody is still asleep). Anyway, we need to see how to migrate over existing series; so need to do some experiments. WDYT?

Yes I wanted to talk to you about that :-) Another potential option would be to land this and then backfill it? I think we could just pause the buildkite pipeline without needing to pause PR submission. If we don't want to miss any intermediate builds on main, we can throw a [block step](https://buildkite.com/docs/pipelines/block-step) in the pipeline and then remove the block step after we've renamed things on the server. I could add the block step in this same PR so there's no timing concerns either. Still would probably be good to do it not on peak hours though.

Experimenting first SGTM. I think we can take a VM instance snapshot of the current server (probably a good idea anyway) and then play around with the Dana UI. Does that sound good? Do you want me to do it, you to do it, or us to do it together?

antiagainst commented:
> Yes I wanted to talk to you about that :-) Another potential option would be to land this and then backfill it? 

I feel backfill might be more difficult than just pause and rewrite.

> I think we could just pause the buildkite pipeline without needing to pause PR submission. If we don't want to miss any intermediate builds on main, we can throw a [block step](https://buildkite.com/docs/pipelines/block-step) in the pipeline and then remove the block step after we've renamed things on the server. I could add the block step in this same PR so there's no timing concerns either. Still would probably be good to do it not on peak hours though.

Yeah. I think it's simpler to do it not in active hours. 

> Experimenting first SGTM. I think we can take a VM instance snapshot of the current server (probably a good idea anyway) and then play around with the Dana UI. Does that sound good? Do you want me to do it, you to do it, or us to do it together?

I think there is no need to do VM level snapshot. The "database" of Dana is just a directory of files. We can just zip that. Let me play with it a bit. I'm happy to do it this time and keep everything documented. So in the future you can handle them! :D

antiagainst reviewed (changes requested):
Could you update TensorFlow models too? We want to convert all of them at once.

GMNGeoffrey commented:
Perfect, thank you! I think adding a block step before the Dana upload in the pipeline would actually be super easy. Then we'll be collecting all the benchmarks but delaying on uploading them

> Could you update TensorFlow models too? We want to convert all of them at once.

I thought we wanted to just abandon those series? Renaming them to say that they came from TFLite source isn't really true. (https://github.com/google/iree/pull/7580 is migrating them all over, but I'm hitting weird vulkan errors)

antiagainst commented:
> > Could you update TensorFlow models too? We want to convert all of them at once.
> 
> I thought we wanted to just abandon those series? Renaming them to say that they came from TFLite source isn't really true. (#7580 is migrating them all over, but I'm hitting weird vulkan errors)

Yup eventually they will be replaced by TFLite ones. But I think that's gonna happen as a next step. (We might also need to first look at the numbers reported by those TFLite ones and resolve obvious issues first before swapping. It's yet to be seen I guess.) It's easier to rename them for now so I can find/replace all at once and don't need to filter stuff. Also it keeps things consistent.

GMNGeoffrey commented:
Ah I was planning to delete them as part of migration to TFLite (that's what #7580 does), assuming the numbers from presubmit look reasonable. Otherwise we're going to be benchmarking three of the models twice in all configurations

GMNGeoffrey commented:
Uh linked to this PR instead of my other one above. Should be #7645 (edited to fix)

antiagainst commented:
> Ah I was planning to delete them as part of migration to TFLite (that's what #7645 does), assuming the numbers from presubmit look reasonable. Otherwise we're going to be benchmarking three of the models twice in all configurations

+1 to avoid benchmarking both the TF and TFLite version: it will certainly blow up the tarball size and uploading/downloading time. 

But I think it's hard to do that switch before this one goes in: we may need to look at the TFLite ones' numbers and fix obvious issues, etc. That could take time. 

Also even after we do the switch, we can still keep the existing Dana series for TF ones for historical references. (They will never be updated after the switch.) It's better to use consistent naming there to avoid confusions.

GMNGeoffrey commented:
> Also even after we do the switch, we can still keep the existing Dana series for TF ones for historical references. (They will never be updated after the switch.) It's better to use consistent naming there to avoid confusions.

Great point about preserving the usability of historical data. I updated all the existing TF benchmarks to be suffixed with "experimental-flags".

I also added a block step to the buildkite pipeline that should prevent any uploads after this PR is submitted. Runs on PRs should be unaffected. When the migration is complete we can remove the block step. All commits between those points will trigger a run of the benchmarks, but not upload them to the dashboard, waiting for someone to unblock. After the migration and removing the block step, we can unblock the pending runs and the upload complete. This is safe assuming that the order in which uploads happen is not material and they each indicate their x-axis value. From reading the code, I believe that is the case.

I didn't really have an easy way to test the block step, so kind of yoloing it, but I believe it's simple enough that we can be relatively confident it will work. I also accidentally didn't limit to the main branch at first, so we can see at least that the block step is well formed on https://buildkite.com/iree/iree-benchmark/builds/1499

ScottTodd commented:
Can we take a backup of the existing benchmark data to have an easy rollback path if we're not happy with how this lands?

GMNGeoffrey commented:
> Can we take a backup of the existing benchmark data to have an easy rollback path if we're not happy with how this lands?

My understanding is that was part of the planned migration. @antiagainst you're going to do that, right?

antiagainst commented:
> I also added a block step to the buildkite pipeline that should prevent any uploads after this PR is submitted.

Nice, thanks!

> This is safe assuming that the order in which uploads happen is not material and they each indicate their x-axis value. From reading the code, I believe that is the case.

Yes, that's true. Order of uploading does not matter.



> > Can we take a backup of the existing benchmark data to have an easy rollback path if we're not happy with how this lands?
> 
> My understanding is that was part of the planned migration. @antiagainst you're going to do that, right?

Yup. I'll save a copy of the existing database (which is just a directory of JSON files really) before updating them.

antiagainst reviewed (approved):
Good to go!
